crates/core/tedge_api/src/workflow/state.rs

```diff
@@ -176,39 +239,86 @@ impl GenericCommandState {
     }
 
     pub fn operation(&self) -> Option<String> {
-        match self.topic.name.split('/').collect::<Vec<&str>>()[..] {
-            [_, _, _, _, _, "cmd", operation, _] => Some(operation.to_string()),
-            _ => None,
-        }
+        extract_command_identifier(&self.topic.name).map(|(operation, _)| operation)
     }
 
     pub fn cmd_id(&self) -> Option<String> {
-        match self.topic.name.split('/').collect::<Vec<&str>>()[..] {
-            [_, _, _, _, _, "cmd", _, cmd_id] => Some(cmd_id.to_string()),
-            _ => None,
-        }
+        extract_command_identifier(&self.topic.name).map(|(_, cmd_id)| cmd_id)
     }
 
-    pub fn is_terminal(&self) -> bool {
-        matches!(self.status.as_str(), "successful" | "failed")
+    pub fn is_successful(&self) -> bool {
+        matches!(self.status.as_str(), SUCCESSFUL)
+    }
+
+    pub fn is_failed(&self) -> bool {
+        matches!(self.status.as_str(), FAILED)
+    }
+}
+
+/// Return the invoking command topic name, if any
+pub fn invoking_command(sub_command: &TopicName) -> Option<TopicName> {
+    match sub_command.split('/').collect::<Vec<&str>>()[..] {
+        [pre, t1, t2, t3, t4, "cmd", _, sub_id] => extract_invoking_command_id(sub_id)
+            .map(|(op, id)| format!("{pre}/{t1}/{t2}/{t3}/{t4}/cmd/{op}/{id}")),
+        _ => None,
```

Comment:
todo: Manual topic parsing without checking topic root

Reply:
Removed by https://github.com/thin-edge/thin-edge.io/pull/2792/commits/da71c7ee5b70f8f21babacc33a1df4eb02efab97

Reply:
Unfortunately, reverted by https://github.com/thin-edge/thin-edge.io/pull/2792/commits/c1450fa94663faa77290690e7bb95645703b5eb4 to fix the regression on "rigger Device Restart Using A Sub-Command"

=> I still have to improve this topic parsing logic

Reply:
Finally addressed by https://github.com/thin-edge/thin-edge.io/pull/2792/commits/8ac28aaec09b9b2d23e92b2b3e80ddaa1f645069